src/network/local_socket_reply.cpp

```diff
@@ -139,38 +143,105 @@ void mp::LocalSocketReply::send_request(const QNetworkRequest& request, QIODevic
+    }
 
     // Build the HTTP User-Agent header
-    http_data += "User-Agent: Multipass/";
-    http_data += mp::version_string;
-    http_data += "\r\n";
+    auto user_agent = request.header(QNetworkRequest::UserAgentHeader).toByteArray();
+    if (!user_agent.isEmpty())
+    {
+        http_data += "User-Agent: " + user_agent + "\r\n";
+    }
+
+    if (!local_socket_write(http_data))
+        return;
 
     if (op == "POST" || op == "PUT")
     {
-        http_data += "Content-Type: application/x-www-form-urlencoded\r\n";
+        http_data = "Content-Type: " + request.header(QNetworkRequest::ContentTypeHeader).toByteArray() + "\r\n";
 
-        if (outgoingData)
+        if (outgoingData && outgoingData->size() > 0)
         {
+            auto content_length = request.header(QNetworkRequest::ContentLengthHeader).toByteArray();
+            auto transfer_encoding = request.rawHeader("Transfer-Encoding");
+            bool is_chunked{false};
+
+            if (!content_length.isEmpty() && !transfer_encoding.isEmpty())
+            {
+                throw mp::HttpLocalSocketException(
+                    "Both \'Content-Length\' and \'Transfer-Encoding\' headers cannot be set at the same time");
+            }
+            else if (content_length.isEmpty() && transfer_encoding.isEmpty())
+            {
+                throw mp::HttpLocalSocketException(
+                    "Either the \'Content-Length\' or the \'Transfer-Encoding\' header must be set");
+            }
+            else if (!content_length.isEmpty())
+            {
+                http_data += "Content-Length: " + content_length + "\r\n";
+            }
+            else
+            {
+                http_data += "Transfer-Encoding: " + transfer_encoding + "\r\n";
+                is_chunked = true;
+            }
+
+            if (!local_socket_write(http_data + "\r\n"))
+                return;
+
+            local_socket->flush();
+
             outgoingData->open(QIODevice::ReadOnly);
+            std::vector<char> data_buffer;
+            data_buffer.reserve(max_bytes);
 
-            http_data += "Content-Length: ";
-            http_data += QByteArray::number(outgoingData->size());
-            http_data += "\r\n\r\n";
-            http_data += outgoingData->readAll();
+            while (true)
+            {
+                auto bytes_read = outgoingData->read(data_buffer.data(), max_bytes);
+                if (bytes_read < 0)
+                {
+                    throw mp::HttpLocalSocketException(
+                        fmt::format("Cannot read data to send to socket: {}", outgoingData->errorString()));
+                }
+
+                if (bytes_read == 0)
+                    break;
+
+                if (is_chunked && !local_socket_write(QByteArray::number(bytes_read, 16) + "\r\n"))
+                    return;
+
+                if (!local_socket_write(QByteArray::fromRawData(data_buffer.data(), bytes_read) + "\r\n"))
```

Comment:
Is trailing `\r\n` expected in non-chunked transfers? How is the server supposed to know it's not part of the data? o.O

I mean, if we exceed `max_bytes` in non-chunked transfers, we'll be putting `\r\n` in the middle of the data?

Reply:
Hmm, yeah, I'll look at this more closely and figure out what the implication is in the case you presented.

Reply:
Fixed